### ISSUE TITLE
Checklist for 6.d

### ISSUE THREAD

**Issue by JJ** [Raku/doc#2632]
## The problem

Many things have changed in the 6.d version, of which we have at least one release. [Here is the list](https://github.com/perl6/roast/blob/master/docs/announce/6.d.md). However, coverage is patchy. Deprecation notices take different form, some new Types and methods are there, some are not... 

Refer to this issue with `refs #2632` and check the item when you work on it either via change in the documentation of after seeing that no changes need to be done to it. In that case, please explain via comment or however why is that the case. 

## Version-Controlled Changes

- [x]  **[6.d]** `&await` no longer blocks *a thread* while waiting
- [x]  **[6.d]** `whenever` not in lexical scope of `react` throws
- [x]  **[6.d]** `$*ARGFILES` inside `sub MAIN` is always fed by `$*IN`
- [x]  **[6.d]** Constructs (literally) `$()`, `@()`, and `%()` are no longer magical
- [x]  **[6.d]** variables with `:D`/`:U` type constraints default to type object of
    the constrained type (e.g. so you could use `.new` with them)
- [x]  **[6.d]** `start` blocks in sink context attach exception handler
~- [ ]  **[6.d]** Routines must use `return-rw` to return a `Proxy`, even if
    routine is marked as `is raw` or `is rw`~
- [x]  **[6.d]** Native `num` types default to `0e0` instead of `NaN`
- [x]  **[6.d]** On subroutine names, the colonpair with key `sym` (e.g. `:sym<foo>`)
    is reserved, in anticipation of possible future use.

## Deprecations

These methods are deprecated in 6.d language and will be removed in 6.e.
Implementations may choose to emit deprecation warnings or to offer these
methods for a longer period than 6.e release.

- [x]  The use of `'-'` (single hyphen) as a special path to `&open` to mean the
    special handles (use `IO::Special` objects instead)
- [x]  `IO::Handle.slurp-rest` (use `.slurp` instead)
- [x]  `Any.flatmap` (use combination of `.flat` and `.map` methods instead)
- [x]  `Cool.path` (use `.IO` instead)
- [x]  `Pair.freeze` (use `Pair.new` with decontainerized arguments instead)
- [x]  `Str.subst-mutate` (use `Str.subst` with `.=` method call assign metaop instead)
- [x]  `Rational.norm` (`Rational` types are required to be normalized on creation now)
- [x]  `IO::Path.child` (use `.add` instead)
- [x]  `&undefine` (assign `Empty`/`Nil` directly, instead)
- [x]  `:count` argument on `&lines`/`Str.lines` routines
    (use `.elems` call on returned `Seq` instead)
- [x]  `&is_approx` in Test.pm6 (use the very similar behaviour of `&is-approx` instead)

## New Behaviors

- [x]  Improved custom handling of `sub MAIN` via new definable `&RUN-MAIN`,
    `&ARGS-TO-CAPTURE`, and `&GENERATE-USAGE` subs
- [x]  `QuantHash`es/`Map` in `%` variables and `List` in `@` variables
    can be declared with `is` trait (e.g. `my %h is Set`)
- [x]  New `<ww>` regex rule: match within word only
- [x]  Loops can produce a list of values from the values of last statements
- [x]  `next`/`last` in a loop that collects its last statement values
    return `Empty` for the iterations they run on
- [x]  `.perl` can be called on consumed `Seq`s, multi-dimensional arrays,
    `Date`, and `CallFrame`
- [x]  `.gist` can be called on `Attribute`
- [x]  Numerous improvements to auto-generated `USAGE` message
- [x]  `is hidden-from-USAGE` trait to hide `sub MAIN` candidates from
    auto-generated `USAGE` message
- [x]  `Parameter.perl` includes introspectable defaults
- [x]  `%*ENV` values are allomorphic
- [x]  Trying to use variables `$;`, `$,`, `$.`, `$\`, `$(`, `$)`,
    `$<`, `$>`, `$/`, `$\`, `$[`, `$-`, `$+`, and `$@`
    throws `X::Syntax::Perl5Var`
- [x]  Default `Hash.keyof` returns a `Str(Any)` coercer type object
- [x]  Non-ASCII numerics can be used in `:42foo` colonpair shortcut
- [x]  `StrDistance` stringifies to its `.after` string
- [x]  More well-defined formatting of Pod tables
- [x]  `Enumeration.enums` returns a `Map`
- [x]  `.Range` on various integer types returns the range of values they support
- [x]  `min`/`max` routines also work on `Hash`es
- [x]  `Signature` literals can contain string/numeric literals as
    well as the invocant marker
- [x]  `List.invert` maps via a required `Pair` binding, resulting in
    potential type check failures
- [x]  `:exists` can be used with multi-dimensional associative subscripts
- [x]  Dynamically created lists can be used to define an enum
- [x]  Junctions can be used as a matcher in `.first`
- [x]  Native attributes can be used as bind targets in parameters
- [x]  `Proc` can work with `IO::Pipe`s from other `Proc`s
- [x]  Typed arrays can be created with both `my SomeType @array`
    and `my @array of SomeType`
- [x]  Items with negative weights are removed when coercing
    a `Mixy` to `Setty`/`Baggy`
- [x]  `:nth` adverb on `m//` accepts a `Junction` as argument
- [x]  `CX::Warn` and `CX::Done` can be caught inside `CONTROL` phaser
- [x]  `next` can be used in `whenever`
- [x]  `require`'d symbols no longer transitively exposed
- [x]  Multi-dimensional access via `{…}`, similar to how it works with `[…]`
- [x]  Any open handles at `END` time get automatically closed
- [x]  On a cached `Seq`, the cached list is used when `&infix:<eqv>`, `.Slip`,
    `.join`, `.List`, `.list`, `.eager`, `.Array` and `.is-lazy` are called
- [x]  `IO::Handle.encoding` takes `Nil` to indicate switch to binary mode
- [x]  `is default` trait works with attributes
- [x]  Parameters with `is rw` trait are considered narrower in multi dispatch
    than those without it
- [x]  `.gist` of `Array`, `Blob`, and `Map` gets trimmed to 100 elements
- [x]  New `for` statement modifiers `hyper for`, `race for`, and `lazy for`
- [x]  `for` loop automatically serializes `RaceSeq`/`HyperSeq`; use new `for`
    statement modifiers `hyper for`/`race for` to avoid
- [x]  `&infix:<does>` can be used with non-composable instances on RHS
- [x]  Numeric comparators can be used with `DateTime` objects
- [x]  `Pod` preserves the type of whitespace
- [x]  Defined semantics for `@`-, `%`- and `&`-sigilled `constant`s

## Math

- [x]  `Rational`s are always reduced on creation and remain
    immutable throughout their life
- [x]  `-Inf`, `Inf`, and `NaN` can be round-tripped through a `Rational` type
    by being represented as values `<-1/0>`, `<1/0>`, and `<0/0`>
    respectively. Zero-denominator `Rational`s are normalized to
    one of those three values
- [x]  Calling `.Int` on ±`Inf` and `NaN` throws
- [x]  Improved IEEE 754-2008 compliance in `Num` operators and math functions
- [x]  Negative zero `Num` (`-0e0`) gets correctly
    handled by all routines and syntactical constructs
- [x]  Stringification of `Num` type is required to be roundtrippable
    to the original `Num`
- [x]  Defined `Complex` exponentiation involving zeros
- [x]  Negative powers in `.expmod` are valid

## Sets, Bags, Mixes (aka QuantHashes) and set operators

- [x]  Set operators can be used on any object, which will be coerced when needed
  - So no pre-coercion is needed or wanted
  - Set operators are at liberty to not create any QuantHash if they can
    perform the desired functionality without them
- [x]  Set operations on different types of QuantHashes will coerce to the
    most liberal form (Set -> Bag -> Mix)
- [x]  The set_precedes family of set operators ( `(<+)`, `≼`, `(>+)`, `≽`) has
    been removed
  - Used to be a Baggy form of the subset operator
  - QuantHashes are upgraded to their most liberal form,
    so `(<=)`, `⊆`, `(>=)`, `⊇` do the right thing
- [x]  `.classify-list` method is available on `Baggy` types
- [x]  `.categorize-list` method is available on `Baggy` types
- [x]  `.invert` method is available on core `QuantHash` types
- [x]  `.antipairs` method can be used on `QuantHash` types
- [x]  `QuantHash` types have `.new-from-pairs` and methods to convert one
    `QuantHash` type to another (e.g. `.Bag` method on `Set` type)
- [x]  `.hash` on `QuantHash` types does stringify keys

## New Parameters and Arguments

- [x]  `Date.new` accepts a `:&formatter`
- [x]  `.first` can take `:kv`
- [x]  `unique` and `.repeated` can take `:&as` and `:&with`
- [x]  `&plan` in Test.pm6 can take `:skip-all`
- [x]  `&run`/`&shell` can take `:merge`
- [x]  `&note` can be called with no arguments
- [x]  `open` accepts `:$out-buffer`
- [x]  `IO::Path.resolve` can take `:completely`
- [x]  `IO::Path.parent` can take an `Int` indicating parent level
- [x]  `Proc::Async.new` slurps positional arguments
- [x]  `Signature.ACCEPTS` accepts non-`Signature`/`Capture` arguments
- [x]  `&EVAL` can take a `Blob`
- [x]  `Promise.keep`/`.break` can be called with no arguments
- [x]  `.sum` on native arrays can take `:wrap`
- [x]  `is required` trait can now take an argument indicating reason
- [x]  `IO::Socket::Async.listen` can bind to port `0` to ask OS for a free port
- [x]  `.encode` can take `:translate-nl`

## New Routines and Operators

- [x]  New `atomicint` Unicode operators and ASCII alternatives that guarantee
    thread-safe, atomic operation:
    `&infix:<⚛=>`/`&atomic-assign`, `&prefix:<⚛>`/`&atomic-fetch`,
    `&prefix:<++⚛>`/`&atomic-inc-fetch`, `&postfix:<⚛++>`/`&atomic-fetch-inc`,
    `&prefix:<--⚛>`/`&atomic-dec-fetch`, `&postfix:<⚛-->`/`&atomic-fetch-dec`,
    `&infix:<⚛-=>`/`&infix:<⚛−=>`/`&atomic-fetch-sub`,
    and `&infix:<⚛+=>`/`&atomic-fetch-add`
- [x]  `&cas`: atomic compare and swap
- [x]  The `≤`, `≥`, and `≠` operators are Unicode operator
    alternatives to `<=`, `>=`, and `!=` respectively
- [x]  `&infix:<unicmp>`/`&infix:<coll>`: alternative behavior of `&infix:<cmp>`
- [x]  `TR///`: non-mutating version of `tr///`
- [x]  `submethod TWEAK`: similar to `BUILD`, except it's compatible with attribute defaults
- [x]  `&duckmap`: apply `&callable` on each element that behaves in such a way
    that `&callable` can be applied
- [x]  `&deepmap`: apply `&callable` on each element, descending into `Iterable`s
- [x]  `&take-rw`: like `&take` but with a writable container
- [x]  `&indir`: execute code in a given `$*CWD`
- [x]  `&spurt`: see `IO::Path.spurt`
- [x]  `&prompt`: prompt user for input
- [x]  `uniprops`: multi-character version of `uniprop`
- [x]  `symlink`: create a file symlink
- [x]  `link`: create a file hardlink
- [x]  `.hyper`/`.race`: process a list of values in parallel
- [x]  `Seq.from-loop`: generate a `Seq` from a `Callable`
- [x]  `Str.uniparse`: parse one or more Unicode character names into
    the actual characters
- [x]  `Str.parse-base`: inverse of `Int.base` operation
- [x]  `IO::Path` provides `.ACCEPTS`, `.SPEC`, `.CWD`, `.Numeric`, `.add`,
    `.extension`, `.mode` and numerious file tests, `.parts`,
    `.sibling`, and `.spurt`
- [x]  `IO::Handle` provides `.READ`, `.WRITE`, `.EOF`, `.DESTROY`,
    `.readchars`, `.flush`, `.lock`, `.unlock`, `.out-buffer`, `.tell`,
    `.say`, `.slurp`, `.seek`, `.printf`, `.print-nl`, and `.watch`
- [x]  `IO::Pipe` provides `.proc`
- [x]  `Iterator` provides `.skip-one`, `.skip-at-least`,
    and `.skip-at-least-pull-one`
- [x]  `Mu.emit`: method form of `&emit`
- [x]  `&fails-like` in Test.pm6 module: allows testing for Failures
- [x]  `&bail-out` in Test.pm6 module: exit out of failing test suite
- [x]  `&is-approx` in Test.pm6 module: test a number is approximately like another
- [x]  `Buf` has `.allocate`, `.reallocate`, `.append`, `.push`, `.pop`,
    `.splice`, `.subbuf-rw`, `.prepend`, and `.unshift` methods
- [x]  `Range` supports `.rand`
- [x]  `Backtrace` has methods `.map`, `.flat`, `.concise`, and `.summary`
- [x]  `.classify-list` method is available on `Hash` types
- [x]  `.categorize-list` method is available on `Hash` types
- [x]  `Code.of`: returns the return type constraint
- [x]  `Code.line`/`.file`: returns the line/file of definition
- [x]  `Proc::Async` provides `.Supply`, `.ready`, `.pid`, `.bind-stdin`,
    `.bind-stdout`, and `.bind-stderr`
- [x]  `Proc.command`/`Proc::Async.command`: the command we're executing
- [x]  `Proc` provides `.signal`, `.pid`, and ~`.encoding`~
- [x]  `Complex` provides `.cis`, `.reals`, `.ceiling`, `.floor`, `.round`,
    `.truncate`, and `.abs` methods and can be compared with `<=>` (as
    long as the imaginary part is negligible)
- [x]  `DateTime` provides `.offset-in-hours`, `.hh-mm-ss`, and `.Date`
- [x]  `DateTime` can be compared with other `DateTime` objects using `<=>` operator
- [x]  `Date` provides `.DateTime` method
- [x]  `&infix:<+>`/`&infix:<->` can be called with `Duration`, `DateTime`,
    and `Real` types
- [x]  `Enumeration` provides `.Int`, `.pred`, `.succ`, `.kv`, and `.pair`
- [x]  `.Date` can be called on an `Instant`
- [x]  Junctions can be created using `Junction.new` call
- [x]  `List` type has `.to` and `.from` methods
- [x]  `Map` type provides `Int` method, returning the number of pairs
- [x]  `Any.skip`: skip values in a list
- [x]  `Any.batch`: more basic cousin of `.rotor`
- [x]  `Mu.iterator`: produce an `Iterator` for values in a list
- [x]  `IO::Spec::*` types provide `.tmpdir`, `.extension`, and `.path`
- [x]  `Pair` provides `.ACCEPTS`, `.Pair`, and `.invert`
- [x]  `.Capture` method is well-defined for all core types
- [x]  Defined semantics of `.ACCEPTS` on allomorphs
- [x]  `Failure.self` explodes unhandled `Failure`s
- [x]  `Thread.is-initial-thread`: are we running in the initial thread?
- [x]  `Match` provides `.Int` and `.actions`
- [x]  `IO::Socket::Async` provides `.socket-port` and `.peer-port`
- [x]  `Promise` provides alternative constructors `.kept` and `.broken`
- [x]  `WhateverCode` provides `.assuming`
- [x]  `WhateverCode` and `Block` provide `.cando`
- [x]  `.:<…>` syntax for calling prefix operators as postfixes
- [x]  `$*KERNEL` provides `.hostname`
- [x]  `Nil` has `.FALLBACK` special method defined to return `Nil`

## New Types

- [x]  `atomicint`: a native `int` sized to be usable with new atomic operators
- [x]  `Lock::Async`: a non-blocking mechanism for mutual exclusion
- [x]  `Encoding::Registry`:  manage available encodings
- [x]  `Encoding::Encoder`: encoder for a specific encoding
- [x]  `Encoding::Decoder`: decoder for a specific encoding
- [x]  `IO::CatHandle`: use multiple read-only `IO::Handle`s as if they were one
- [x]  Native `str` arrays
- [x]  `Supplier::Preserving`: cached live `Supply` factory
- [x]  `Semaphore`: control access to shared resources by multiple threads
- [x]  `IO::Special`: a path to special I/O device (e.g. `STDOUT`)
- [x]  `Exceptions::JSON` an implementation of custom exceptions handler
    (can be used with `PERL6_EXCEPTIONS_HANDLER` env var)
- [x]  `SeekType` enum: values for use in `IO::Handle.seek`

## New Variables

- [x]  `$*USAGE`: available inside `MAIN` subs and contains the auto-generated
  USAGE message
- [x]  `%*SUB-MAIN-OPTS`: settings for behaviour of `sub MAIN`
    - `%*SUB-MAIN-OPTS<named-anywhere>` allow named arguments to be
        placed at any position on the command line
- [x]  `$*COLLATION`: configures the four Unicode collation levels
- [x]  `$*INIT-INSTANT`: an `Instant` representing program startup time
- [x]  `$*HOME`: user's home directory, if one exists
- [x]  `&*chdir`: a `Callable` containing a variant of `IO::Path.chdir` that
    also sets process's current directory
- [x]  `PERL6_TEST_DIE_ON_FAIL` environmental variable: stop test
    suite on first failure
- [x]  `PERL6_EXCEPTIONS_HANDLER` environmental variable: specify custom
    exceptions handler class

## Clarifications of Edge Case/Coercion Behaviour

- [x]  `UInt` smartmatches `True` with `Int` type object
- [x]  `sink` statement prefix explodes `Failure`s
- [x]  Defined behaviour of `permutations`/`combinations` on 1- and 0-item
    lists and negative and non-Int arguments
- [x]  `&val`, `Str.Numeric`, and other `Str` numeric conversion methods
    `fail` when trying to convert Unicode `No` character
    group or synthetic numerics
- [x]  Synthetic numerics cannot be used in `:42foo` colonpair shortcut
- [x]  An `Enumeration` can now be used as a array shape specifier
- [x]  Numeric conversion of `Str` containing nothing but whitespace returns `0` now
- [x]  `samemark` with empty pattern argument simply returns the invocant
- [x]  `.polymod` can be used with `lazy` but finite lists of divisors
- [x]  `.[*-0]` index is defined
- [x]  Negative gaps in `.rotor` that are larger than the sublist throw
- [x]  Non-`Int` arguments to `.rotor` get coerced to `Int`
- [x]  `.lines` is defined when reading `/proc` files
- [x]  Defined behaviour of Thai numerals in postfix/prefix `++`/`--` on strings
- [x]  `map` inside sunk `for` is treated as sunk
- [x]  Sunk `for` loop sinks value of last statement's method call
- [x]  `.Int` on `Bool` objects returns an `Int` object
- [x]  `splice` can be used to extend an array
- [x]  `classify` works with `Junction`s
- [x]  `.pairup` on a type object returns an empty `Seq`
- [x]  `.pairup` always returns a `Seq`
- [x]  Synthetic codepoints are rejected from `Date`/`DateTime` constructors
- [x]  `⸨`/`⸩` pair can now be used as matching characters in quoting constructs
- [x]  `.flat` on an `Array` type object simply returns that type object
- [x]  Mixed-level `classify` on `Hash`es throws
- [x]  `Junction`s can be used to specify multiple keys to `Hash`es
- [x]  The `Callable` given to `.classify-list` is now guaranteed to be
    executed only once per each item
- [x]  `:delete` on associative lookup on `Hash` type object returns `Nil`
- [x]  `&is-deeply` from Test.pm6 automatically `.cache`s `Seq`s given
    as arguments and uses the returned `List`s for testing
- [x]  `Complex.new()` gives `<0+0i>`
- [x]  `Int.new` is now guarantied to construct a new `Int` (rather than, say,
       re-use one from a constants cache)
- [x]  1-arg versions of `&infix:<=:=>` and `&infix:<eqv>` are defined
- [x]  `Nil` type now throws if directly or indirectly calling
    `.BIND-POS`, `.BIND-KEY`, `.ASSIGN-POS`, `.ASSIGN-KEY`, `.STORE`, `.push`,
    `.append`, `.unshift`, `.prepend`
- [x]  `Nil.ord` returns an empty `Seq`
- [x]  `Nil.chrs` returns a `"\0"`
- [x]  `Num.new` coercers argument to `Num`
- [x]  `infix:<Z>()` returns an empty `Seq`
- [x]  `.comb` always returns a `Seq`
- [x]  Reduce with `&infix:<+>` with one item simply returns that item
- [x]  `()[0]` returns `Nil`
- [x]  Regex smartmatching is allowed on (possibly-infinite) `Seq`
- [x]  Defined smartmatching with `Range` objects
- [x]  `Set` converted to a `Mix`/`Bag` no longer has `Bool` weights
- [x]  `gcd` is defined when one or more operands are `0`
- [x]  `Junction`s autothread in `defined` routine
- [x]  `sum` can handle lists with `Junctions` in them
- [x]  `Grammar.parse` lets top `regex` backtrack
- [x]  The `U+2212 MINUS SIGN [Sm] (−)` is now supported by more constructs,
    such as `Str.Numeric` and `&val`
- [x]  Arity-1 `&infix:<~>` works with `Blob`s
- [x]  All `Numeric` literals are supported as value literals in signature
- [x]  `\b` and `\B` in regexes throw `X::Obsolete`
- [x]  `True` and `False` as value literals in signatures warn
- [x]  Return type of `.sort` and `IO::Spec::Unix.path` is always `Seq`
- [x]  Out-of-range `.AT-POS` on `Range` objects returns `Nil`
- [x]  `Pair.AT-KEY` for non-existent key returns `Nil`
- [x]  All `Cool` types provide `.Rat`/`.FatRat` coercers
- [x]  `IO::Path` filetests do not cache results of earlier test executions
- [x]  `Seq` eqv `List` as `False` based on type mismatch alone
~- [ ]  On arrays, `Hash`es, and `QuantHash`es, values from `.kv`, `.values`, and `.pairs` sequences are writable~ See perl6/roast#614 and #3519 
- [x]  `&infix:<∘>`/`&infix:<o>` keep LHF's `.of` and RHS's `.arity` and `.count`
- [x]  Refined accepted arguments in regex operator adverbs (e.g. `:in(…)`)
- [x]  Refined accepted combinations of arguments in `IO::Handle.open`
- [x]  `IO::Path.Str` does not include the value of `.CWD` attribute
- [x]  `IO::Path` type rejects paths with the `nul` byte (`"\0"`) in them
- [x]  `IO::Pipe`'s `.path`/`.IO` return an `IO::Path` type object
- [x]  `IO::Path`'s `.copy`/`.move` `fail` if destination and original are the same
- [x]  `dir`-created `IO::Path`s' absoluteness is controlled by the invocant
- [x]  More-defined edge-case behaviour, `Callable` handling, `.defined` calling,
    and chaining of `&infix:<andthen>`, `&infix:<orelse>`,
    and `&infix:<notandthen>` operators
- [x]  Zen slicing of `Seq`s does *not* cache them
- [x]  `List.Capture` stringifies keys of any contained `Pair` objects
- [x]  `&fail` with handled `Failure` argument marks it as unhandled
- [x]  `use lib` accepts `IO::Path` objects
- [x]  Anchors `^`, `^^`, `$`, and `$$` are valid in lookarounds
- [x]  `Grammar.made` supports type objects
- [x]  `.isa` supports `subset` type objects
- [x]  `:delete` can be used on lazy arrays
- [x]  `&infix:<eqv>` can work with certain cases of lazy arguments
- [x]  Dynamic lookup (`::(…)`) is restricted regex syntax and
    requires `use MONKEY-SEE-NO-EVAL` clearance
- [x]  Defined `.Slip` and `.List` on arrays with holes
- [x]  `Promise.in`/`.at` and `Supply.interval` work with zero and negative values
- [x]  `Supply.interval` minimum value is `0.001`; lower values are
    treated as `0.001` and emit warnings # mentioned at https://docs.perl6.org/type/Supply#method_interval PR https://github.com/perl6/doc/pull/2649
- [x]  `Supply` provides `.Seq`, `.list`, and `.zip`
- [x]  Can bind to native type attributes in build methods
- [x]  `WhateverCode` propagates `use fatal`
- [ ]  `say`, `note`, `put`, `print`, and `printf` routines autothread `Junction`s
- [x]  `IO::Handle.eof` value changes accordingly when `.seek`ing past end and back
- [x]  Defined `.succ`, `.pred`, and `.Bool` on allomorphs
- [x]  Defined `.Bridge` on core `Numeric`s
~- [ ]  Defined `.Numeric`/`.Real` on type objects of core `Numeric`s~
- [x]  Defined `Rational.Bool` with respect to zero-denominator rationals
- [x]  `say`/`note` guaranteed to call `.gist` on subclasses of `Str`
- [x]  Defined `Junction.Str` returns a `Junction`
- [x]  Defined `Junction.gist`/`.perl` return a `Str`
- [x]  `Map`/`Hash`'s `.list`/`.cache` return a `List`
- [x]  Defined `.round`'s return type
- [x]  Defined `Enumeration:D` not does `.ACCEPT` an `Enumeration:U` See rakudo/rakudo#2073

## Miscellaneous

- [x]  The `IO::ArgFiles` type is just an empty subclass of `IO::CatHandle`
- [x]  Constraints on constants
    - Constraints are fully enforced
    - Attempting to use parametarized type constraints on constants
        (i.e. using `my Foo constant @int`) throws `X::ParametricConstant` exception
- [x]  `Pod` `=defn` (definition list) directive is available
- [x]  `Pod` provides `:numbered` config key
- [x]  `.^ver`, `.^auth`, and `.^name` metamethods are available on `module`
    and are absent on a `package`, by design
- [x]  Fancy quotes (`’…’`, `“…”`, `｢…｣`, and variants) are supported in `qww<…>`
- [x]  `&infix:< >` supports lookup of autogenerated `Callables` (e.g. `&infix:<XX>`)
- [x]  Using a named `anon` sub no longer produces redeclaration warnings
- [x]  Extended spec of `::?MODULE`/`$?MODULE` variable
- [x]  `sub MAIN` can accept an `Enumeration` type constraint and `where`
    clause on arguments
- [x]  Type smiley constraints can be used on subsets
- [x]  `start` blocks and thunks get fresh `$/` and `$!`
- [x]  `R` meta operator used with list-associative operators is defined
- [x]  Type coerces can be used in signature return type constraints
- [x]  `&infix:<x>`/`&infix:<x>` throw with `-Inf`/`NaN` repeat arguments
- [x]  Literal constructs `put` and `put for` throw, requiring use of parentheses
- [x]  Expanded specification coverage of Unicode routines and features - Upgraded coverage to Unicode version 11
- [x]  `$.` method call syntax shorthand works with meta-methods

**Comment by JJ** [comment]
I'm going to be starting with deprecations. Please work on the rest, and say here what you will do.

**Comment by JJ** [comment]
OK; I'll go now for the first section, version-controlled changes.

**Comment by JJ** [comment]
Can someone clarify what:

> `whenever` not in lexical scope of `react` throws

means?

Does it mean in supply *or* react scope?

**Comment by AlexDaniel** [comment]
@JJ yes, previously this was allowed:
```
sub foo() {
    whenever Promise.in(2) { say ‘hello world’ }
}
react {
    foo();
}
```
Notice how `whenever` is not in the scope of the `react` block. This used to work previously, but it won't anymore. Now you're required to write:
```
react {
    whenever Promise.in(2) { say ‘hello world’ }
}
```

**Comment by JJ** [comment]
Can anyone tell me what this:

>  Constructs (literally) `$()`, `@()`, and `%()` are no longer magical

means? Or what it means that _before_ they were magical?

**Comment by Leont** [comment]
> Can anyone tell me what this means

See [this](https://github.com/rakudo/rakudo/issues/1946) issue

**Comment by JJ** [comment]
In the case of C<Proxy>s
> Routines must use return-rw to return a Proxy, even if routine is marked as is raw or is rw

I have checked the example, which does not use return-rw, and seems to work without a problem. return-rw works in the same way. Doesn't *must* imply it's going to fail otherwise?

**Comment by JJ** [comment]
I think that the spec above wasn't finally done [due to this](https://github.com/rakudo/rakudo/issues/2113)

**Comment by JJ** [comment]
I don't think this is true:
>  variables with :D/:U type constraints default to type object of the constrained type (e.g. so you could use .new with them)

There's no way to get the default value. So even if that's the case, I don't think it should go to the documentation.

**Comment by antoniogamiz** [comment]
I'm currently working at Math section :D! I'll update here my progress.

**Comment by AlexDaniel** [comment]
> I think that the spec above wasn't finally done due to this

No, not really. You *must* use return-rw or is-rw. We added bug compatibility so that some rare cases of existing code keep working, but nobody should expect that to work and we didn't make any guarantees on how long it will keep working.

**Comment by ElenaMerelo** [comment]
Hey! I am going to see whether .perl can be called on consumed Seqs, multi-dimensional arrays,
Date, and CallFrame :)

**Comment by ElenaMerelo** [comment]
Done that, now i am going to add the fact that Parameter.perl includes introspectable defaults

**Comment by ElenaMerelo** [comment]
That is already being worked on, hence i am going to add that Trying to use variables $;, $,, $., $\, $(, $),
$<, $>, $/, $\, $[, $-, $+, and $@
throws X::Syntax::Perl5Var

**Comment by JJ** [comment]
@ElenaMerelo I'm kind of baffled by that. I think the best option is to either open the question to see if anyone else knows what it means, or add it verbatim to the introduction to Parameter.

**Comment by ElenaMerelo** [comment]
I am leaving the point **Trying to use variables $;, $,, $., $\, $(, $),
$<, $>, $/, $\, $[, $-, $+, and $@** since they are not documented.

**Comment by Altai-man** [comment]
I am looking at `New variables` section...

**Comment by Altai-man** [comment]
>min/max routines also work on Hashes
Judging by https://colabti.org/irclogger/irclogger_log/perl6?date=2019-03-03#l253 both method and routine forms have worked in 6.c too. Those are not documented explicitly(`Hash` page does not have it among other methods). But `min` and `max` are defined on `Any` which coerces the invocant to Iterable, and given that Hash ~~ Iterable, it just works.

I don't know what should be done in this situation.

**Comment by Altai-man** [comment]
What is the state of docs versioning, by the way? Spec version-wise.
A lot of things are hard to fix here, because one has to write something like "As for 6.d spec version, X is now Y", which is not very robust.

I have also read about docs site re-implementation efforts, @finanalyst ? Is there anything I can help with?

**Comment by JJ** [comment]
El lun., 4 mar. 2019 a las 8:56, Altai-man (<notifications@github.com>)
escribió:

> What is the state of docs versioning, by the way? Spec version-wise.
>
Chaotic.

> A lot of things are hard to fix here, because one has to write something
> like "As for 6.d spec version, X is now Y", which is not very robust.
>
> I have also read about docs site re-implementation efforts, @finanalyst
> <https://github.com/finanalyst> ? Is there anything I can help with?
>

There's the document for reimplementation
https://github.com/perl6/doc/wiki/A-Redesign-of-the-Perl-6-Documentation-System,
which includes a reformulation of the document meta-data. It's also been
included as an idea for GSoC. It would be great if you volunteered as a
mentor.

**Comment by JJ** [comment]
As a matter of fact, this #2647 goes in that direction. It would be interesting if we could at least try to make document releases close to the Rakudo releases. If we managed to close this issue, we could try to ocntinue with the monthly (or bimonthly, as of now) releases.

**Comment by antoniogamiz** [comment]
As for the item:
Trying to use variables $;, $,, $., $\, $(, $), $<, $>, $/, $\, $[, $-, $+, and $@ throws X::Syntax::Perl5Var, tobs says that according to Grammar.nqp that exception is only if you try to assign to $, , for example. 

In fact, in some cases ($\), it throws the exception when you use it (like say $\;).

**Comment by JJ** [comment]
@antoniogamiz Will you include that in the docs?

**Comment by lukasvalle** [comment]
For this changelog I can't get, what has changed for 6.d:

> .Int on Bool objects returns an Int object


    say $*PERL.compiler;              #rakudo (2018.04.1)
    say Bool.new( False ).Int ~~ Int  #True


    use v6.c; 
    say $*PERL.compiler;              #rakudo (2019.03.1)
    say Bool.new( False ).Int ~~ Int  #True


    say $*PERL.compiler;              #rakudo (2019.03.1)
    say Bool.new( False ).Int ~~ Int  #True

**Comment by JJ** [comment]
It might refer to a new implementation, instead of the default
implementation that it used before. Will have to be checked a bit
further...

**Comment by lukasvalle** [comment]
> The ≤, ≥, and ≠ operators are Unicode operator alternatives to <=, >=, and != respectively

Are in the operators docs:
[≤](https://docs.perl6.org/language/operators#infix_%E2%89%A4)
[≥](https://docs.perl6.org/language/operators#infix_%E2%89%A5)
[≠](https://docs.perl6.org/language/operators#infix_%E2%89%A0)

mark the item as done?

**Comment by Altai-man** [comment]
@lukasvalle it is not necessarily a change. The changelog describes changes in the specification. Some things have worked before this way in Rakudo, but it was not specified in the spec itself. 6.d made the thing explicit instead of "Oh, we have a compiler X that does it this way, but the spec says nothing". If it is mentioned somewhere, it is enough for such an item.

**Comment by Altai-man** [comment]
@lukasvalle yes, if the thing is documented already, just mark the item as done.

**Comment by lizmat** [comment]
"Parameter.perl includes introspectable defaults" is a non-issue.  We don't document `.perl` methods in general, so we don't need to here either, I would say.  What changed is that the output of `Parameter.perl` is now much more complete, and that you can subclass `Parameter`s better than before.  Which is already documented, afaics.

**Comment by rafaelschipiura** [comment]
> Hash.new.of
(Mu)

Is this the one supposed to be Str(Any)?

> say $*PERL.compiler.version
v2018.12
> say $*PERL
Perl 6 (6.d)

**Comment by lizmat** [comment]
No, `Hash.new.keyof` is returns the `Str(Any` coercion type.

**Comment by lukasvalle** [comment]
.gist of Array, Blob, and Map gets trimmed to 100 elements:
-> Array: is referenced to List and documented there
-> Blob: in Blob documentation
-> Map: in Map documentation

**Comment by JJ** [comment]
Just check it out, then. In many cases it might be already done.

**Comment by lukasvalle** [comment]
'New for statement modifiers hyper for, race for, and lazy for'
lazy for: already in /syntax/lazy
race for and hyper for: added with commit dc7f8a70c3beab85c2163290bf63b66a369ef3c3

**Comment by lukasvalle** [comment]
' .antipairs method can be used on QuantHash types'
according to the type graph, this are Setty and Baggy
- Baggy: already there
- Setty: added with commit a7c427895d0993fe403408a2a4f5b8490fe61871

**Comment by lukasvalle** [comment]
'IO::Path provides .ACCEPTS, .SPEC, .CWD, .Numeric, .add, .extension, .mode and numerious file tests, .parts, .sibling, and .spurt'
all there in /type/IO::Path

**Comment by lukasvalle** [comment]
'IO::Handle provides .READ, .WRITE, .EOF, .DESTROY, .readchars, .flush, .lock, .unlock, .out-buffer, .tell, .say, .slurp, .seek, .printf, .print-nl, and .watch'
All there expect .watch. But .watch is also not implemented for IO::Handle. It is for IO::Path. Marked as done.

**Comment by lukasvalle** [comment]
'Iterator provides .skip-one, .skip-at-least, and .skip-at-least-pull-one'
- all there in /type/Iterator

**Comment by lukasvalle** [comment]
IO::Path's .copy/.move fail if destination and original are the same
- fail for both routines already documented --> marked completed

**Comment by lukasvalle** [comment]
IO::Pipe's .path/.IO return an IO::Path type object
- return type for both routines already documented --> marked completed

**Comment by lukasvalle** [comment]
IO::Path.Str does not include the value of .CWD attribute
- already documented --> marked completed

**Comment by lukasvalle** [comment]
PERL6_TEST_DIE_ON_FAIL there: language/testing
PERL6_EXCEPTIONS_HANDLER there: programs/03-environment-variables

**Comment by lukasvalle** [comment]
".encode can take :translate-nl" there in routine/encode

**Comment by lukasvalle** [comment]
" IO::Special: a path to special I/O device (e.g. STDOUT)" there in type/IO::Special

**Comment by lukasvalle** [comment]
"SeekType enum: values for use in IO::Handle.seek" listed at type/IO::Handle#method_seek

**Comment by lukasvalle** [comment]
" .pairup always returns a Seq" return type Seq mentioned at: routine/pairup

**Comment by lukasvalle** [comment]
".pairup on a type object returns an empty Seq": PR https://github.com/perl6/doc/pull/2845

**Comment by lukasvalle** [comment]
Map/Hash's .list/.cache return a List:  
- .list at: type/Map#method_list and inherited by Hash
-.cache at: routine/cache

**Comment by lukasvalle** [comment]
"List.invert maps via a required Pair binding, resulting in potential type check failures" mentioned at: routine/invert#class_List

**Comment by lukasvalle** [comment]
Defined smartmatching with Range objects is there at: type/Range#method_ACCEPTS

**Comment by lukasvalle** [comment]
"Defined semantics of .ACCEPTS on allomorphs". .ACCEPTS already documented for:
- IntStr, NumStr, RatStr and ComplexStr

**Comment by lukasvalle** [comment]
> &*chdir: a Callable containing a variant of IO::Path.chdir that also sets process's current directory 

there at routine/&*chdir

**Comment by lukasvalle** [comment]
"Exceptions::JSON an implementation of custom exceptions handler" documented at: https://docs.perl6.org/programs/03-environment-variables#Error_message_verbosity_and_strictness

**Comment by JJ** [comment]
I'm going to mark `&infix:<does>` as done, after spinning off #3131.

**Comment by JJ** [comment]
Pod non-preserving whitespace was a bug and was fixed [here](https://github.com/rakudo/rakudo/issues/1852)

**Comment by JJ** [comment]
Created perl6/roast#614 to test writable kv, pairs, all that. Wait for tests before adding it to the documentation

**Comment by lukasvalle** [comment]
> Literal constructs put and put for throw, requiring use of parentheses

is documented on the independent-routines page [put](https://docs.raku.org/language/independent-routines#sub_put)

**Comment by lukasvalle** [comment]
> Promise.in/.at and Supply.interval work with zero and negative values

Promise.in already mentions that negative values are treated as 0
Promise.at already mentions that time is in the past, it will be treated as now
Supply.interval mention negative values added with https://github.com/Raku/doc/pull/3358

**Comment by lukasvalle** [comment]
> WhateverCode propagates use fatal

References very likely to https://github.com/rakudo/rakudo/commit/31db3accf4 which is a fix and not a new behavior. No need to modify doc.

**Comment by JJ** [comment]
> Grammar.made supports type objects

This does not make a lot of sense. made is defined in NQPMatchRole and redefined in Match; there's no override in Grammar. I'll try and find out what it means.

**Comment by JJ** [comment]
It's likely it's [this commit](https://github.com/rakudo/rakudo/commit/d0d105b8b6201c1a2d11f2645d386981a2ea806a). In that case, it's `make`, not `made`

**Comment by lukasvalle** [comment]
_IO::Handle.eof value changes accordingly when .seeking past end and back_ documented in [IO:: Handle eof](https://docs.raku.org/type/IO::Handle#method_eof)

_Type coerces can be used in signature return type constraints_ documented in [Signature Coercion_type](https://docs.raku.org/type/Signature#Coercion_type)

**Comment by JJ** [comment]
I couldn't find what "Defined `.Numeric`/`.Real` on type objects of core `Numeric`s" means, scratch it out.

**Comment by JJ** [comment]
The
> Dynamic lookup (::(…)) is restricted regex syntax and
requires use MONKEY-SEE-NO-EVAL clearance

item refers to [this PR](https://github.com/rakudo/rakudo/pull/1168). Trying to understand what "restricted syntax" means and what it really implies.
